File: pvm/opcode_handlers9.go
package pvm

import "math/bits"

// A.5.9. Instructions with Arguments of Two Registers.

func handleMOVE_REG(vm *VM, operands []byte) {
	// Inline register extraction with minimal bounds checking
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)

	// Use branchless bounds check with bitwise operations
	// Since indices are 4-bit (0-15), we only need to clamp values > 12
	if registerIndexD > 12 {
		registerIndexD = 12
	}
	if registerIndexA > 12 {
		registerIndexA = 12
	}

	vm.register[registerIndexD] = vm.register[registerIndexA]
	if PvmTrace {
		dumpMov(registerIndexD, registerIndexA, vm.register[registerIndexA])
	}
	vm.pc += 1 + uint64(len(operands))
}

func handleSBRK(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]

	if valueA == 0 {
		vm.register[registerIndexD] = uint64(vm.GetCurrentHeapPointer())
		vm.pc += 1 + uint64(len(operands))
		return
	}

	result := uint64(vm.GetCurrentHeapPointer())
	next_page_boundary := P_func(vm.GetCurrentHeapPointer())
	new_heap_pointer := uint64(vm.GetCurrentHeapPointer()) + valueA

	if new_heap_pointer > uint64(next_page_boundary) {
		final_boundary := P_func(uint32(new_heap_pointer))
		idx_start := next_page_boundary / Z_P
		idx_end := final_boundary / Z_P
		page_count := idx_end - idx_start

		vm.allocatePages(idx_start, page_count)
	}
	vm.SetCurrentHeapPointer(uint32(new_heap_pointer))
	if PvmTrace {
		dumpTwoRegs("SBRK", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleCOUNT_SET_BITS_64(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(bits.OnesCount64(valueA))
	if PvmTrace {
		dumpTwoRegs("COUNT_SET_BITS_64", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleCOUNT_SET_BITS_32(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(bits.OnesCount32(uint32(valueA)))
	if PvmTrace {
		dumpTwoRegs("COUNT_SET_BITS_32", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleLEADING_ZERO_BITS_64(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(bits.LeadingZeros64(valueA))
	if PvmTrace {
		dumpTwoRegs("LEADING_ZERO_BITS_64", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleLEADING_ZERO_BITS_32(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(bits.LeadingZeros32(uint32(valueA)))
	if PvmTrace {
		dumpTwoRegs("LEADING_ZERO_BITS_32", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleTRAILING_ZERO_BITS_64(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(bits.TrailingZeros64(valueA))
	if PvmTrace {
		dumpTwoRegs("TRAILING_ZERO_BITS_64", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleTRAILING_ZERO_BITS_32(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(bits.TrailingZeros32(uint32(valueA)))
	if PvmTrace {
		dumpTwoRegs("TRAILING_ZERO_BITS_32", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleSIGN_EXTEND_8(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(int8(valueA & 0xFF))
	if PvmTrace {
		dumpTwoRegs("SIGN_EXTEND_8", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleSIGN_EXTEND_16(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := uint64(int16(valueA & 0xFFFF))
	if PvmTrace {
		dumpTwoRegs("SIGN_EXTEND_16", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleZERO_EXTEND_16(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := valueA & 0xFFFF
	if PvmTrace {
		dumpTwoRegs("ZERO_EXTEND_16", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

func handleREVERSE_BYTES(vm *VM, operands []byte) {
	// Inline register extraction
	regByte := operands[0]
	registerIndexD := int(regByte & 0x0F)
	registerIndexA := int(regByte >> 4)
	if registerIndexD > 12 { registerIndexD = 12 }
	if registerIndexA > 12 { registerIndexA = 12 }
	
	valueA := vm.register[registerIndexA]
	result := bits.ReverseBytes64(valueA)
	if PvmTrace {
		dumpTwoRegs("REVERSE_BYTES", registerIndexD, registerIndexA, valueA, result)
	}
	vm.register[registerIndexD] = result
	vm.pc += 1 + uint64(len(operands))
}

File: pvm/src/opcode_handlers9.c
#include "pvm.h"
#include <inttypes.h>


// Utility functions
extern uint32_t p_func(uint32_t x);

// Bit manipulation functions (C equivalents of Go's math/bits)
static inline int count_set_bits_64(uint64_t x) {
    return __builtin_popcountll(x);
}

static inline int count_set_bits_32(uint32_t x) {
    return __builtin_popcount(x);
}

static inline int leading_zero_bits_64(uint64_t x) {
    return x ? __builtin_clzll(x) : 64;
}

static inline int leading_zero_bits_32(uint32_t x) {
    return x ? __builtin_clz(x) : 32;
}

static inline int trailing_zero_bits_64(uint64_t x) {
    return x ? __builtin_ctzll(x) : 64;
}

static inline int trailing_zero_bits_32(uint32_t x) {
    return x ? __builtin_ctz(x) : 32;
}

static inline uint64_t reverse_bytes_64(uint64_t x) {
    return __builtin_bswap64(x);
}

// A.5.9. Instructions with Arguments of Two Registers

void handle_MOVE_REG(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction with minimal bounds checking
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    
    // Use branchless bounds check with bitwise operations
    // Since indices are 4-bit (0-15), we only need to clamp values > 12
    if (register_index_d > 12) {
        register_index_d = 12;
    }
    if (register_index_a > 12) {
        register_index_a = 12;
    }
    
    vm->registers[register_index_d] = vm->registers[register_index_a];
    
    vm->pc += 1 + operand_len;
}

void handle_SBRK(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    uint64_t value_a = vm->registers[register_index_a];
    
    if (value_a == 0) {
        vm->registers[register_index_d] = (uint64_t)pvm_get_current_heap_pointer(vm);
        vm->pc += 1 + operand_len;
        return;
    }

    uint64_t result = (uint64_t)pvm_get_current_heap_pointer(vm);
    uint32_t next_page_boundary = p_func(pvm_get_current_heap_pointer(vm));
    uint64_t new_heap_pointer = (uint64_t)pvm_get_current_heap_pointer(vm) + value_a;

    if (new_heap_pointer > (uint64_t)next_page_boundary) {
        if ( vm->pvm_tracing ) {
            printf("TRACE SBRK: new_heap_pointer=0x%llx next_page_boundary=0x%x\n",
                   (unsigned long long)new_heap_pointer, next_page_boundary);
            fflush(stdout);
        }
        uint32_t final_boundary = p_func((uint32_t)new_heap_pointer);
        uint32_t idx_start = next_page_boundary / Z_P;
        uint32_t idx_end = final_boundary / Z_P;
        uint32_t page_count = idx_end - idx_start;
        pvm_allocate_pages(vm, idx_start, page_count);
    }
    pvm_set_current_heap_pointer(vm, (uint32_t)new_heap_pointer);
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_COUNT_SET_BITS_64(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)count_set_bits_64(value_a);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_COUNT_SET_BITS_32(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)count_set_bits_32((uint32_t)value_a);
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_LEADING_ZERO_BITS_64(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)leading_zero_bits_64(value_a);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_LEADING_ZERO_BITS_32(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)leading_zero_bits_32((uint32_t)value_a);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_TRAILING_ZERO_BITS_64(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)trailing_zero_bits_64(value_a);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_TRAILING_ZERO_BITS_32(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)trailing_zero_bits_32((uint32_t)value_a);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_SIGN_EXTEND_8(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)(int8_t)(value_a & 0xFF);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_SIGN_EXTEND_16(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = (uint64_t)(int16_t)(value_a & 0xFFFF);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_ZERO_EXTEND_16(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = value_a & 0xFFFF;
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}

void handle_REVERSE_BYTES(VM* vm, uint8_t* operands, size_t operand_len) {
    // Inline register extraction
    uint8_t reg_byte = operands[0];
    int register_index_d = (int)(reg_byte & 0x0F);
    int register_index_a = (int)(reg_byte >> 4);
    if (register_index_d > 12) register_index_d = 12;
    if (register_index_a > 12) register_index_a = 12;
    
    uint64_t value_a = vm->registers[register_index_a];
    uint64_t result = reverse_bytes_64(value_a);
    
    
    vm->registers[register_index_d] = result;
    vm->pc += 1 + operand_len;
}


